### PR TITLE
Smos l4

### PR DESCRIPTION
**Pull request by Adeaem**
Readers, writers, image to time-series processor and tests for scientific and operational CATDS-CESBIO SMOS L4 Root-zone soil moisture products

### CODE REVIEW

File: src/smos/smos_l4/interface.py [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+        timestamps = []
+        diff = end_date - start_date
+        for i in range(diff.days + 1):
+            daily_dates = start_date + timedelta(days=i) + img_offsets
+            timestamps.extend(daily_dates.tolist())
+
+        return timestamps
+
+
+class SMOSL4Ts(GriddedNcOrthoMultiTs):
```
Comment (wpreimes) [inline src/smos/smos_l4/interface.py:478]: Write a common time series reader for all SMOS Ts products

File: src/smos/smos_l4/interface.py [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+from netCDF4 import Dataset, date2num, num2date
+from pygeogrids.netcdf import load_grid
+from smos.grid import EASE25CellGrid
+from smos.smos_ic import dist_name as subdist_name
+from smos import dist_name, __version__
+from collections import OrderedDict
+import pandas as pd
+
+
+class SMOSL4Img(ImageBase):
```
Comment (wpreimes) [inline src/smos/smos_l4/interface.py:41]: a lot of duplicate code here as well, maybe we should make a common parent class that the smos-ic and l4 `Image` and `Dataset` classes then inherit from? And adapt it according to each dataset. I think that would make the whole package a bit more slim and easier to maintain, resp add additional products.

**Review by wpreimes — CHANGES_REQUESTED** [review]

File: tatus [hunk 3]
```diff
@@ -0,0 +1,244 @@
+[33mcommit 65920c25a810d1e8d0684973a02f9a4027203a50[m[33m ([m[1;36mHEAD -> [m[1;32msmos_l4[m[33m)[m
```
[the remark this answers is not in the document]
Comment (Adeaem) [reply tatus]: removed

File: src/smos/smos_l4/reshuffle.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+import argparse
+from datetime import datetime
+
+from repurpose.img2ts import Img2Ts
+from smos.smos_l4.interface import SMOSL4Img, SMOSL4Ds
+from smos.grid import EASE25CellGrid
+from netCDF4 import Dataset
+
+
+def firstfile(input_root):
```
[the remark this answers is not in the document]
Comment (Adeaem) [reply src/smos/smos_l4/reshuffle.py]: done

File: src/smos/smos_l4/reshuffle.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                firstfile = os.path.join(input_root, curr, f)
+                ds = Dataset(firstfile)
+                vars = ds.variables.keys()
+
+                return firstfile, vars
+
+    raise (IOError, 'No netcdf data found under the passed root directory')
+
+
+def mkdate(datestring):
```
[the remark this answers is not in the document]
Comment (Adeaem) [reply src/smos/smos_l4/reshuffle.py]: done

File: src/smos/smos_l4/reshuffle.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    datestr : datetime
+        Date string as datetime.
+    """
+    if len(datestring) == 10:
+        return datetime.strptime(datestring, '%Y-%m-%d')
+    if len(datestring) == 16:
+        return datetime.strptime(datestring, '%Y-%m-%dT%H:%M')
+
+
+def str2bool(val):
```
[the remark this answers is not in the document]
Comment (Adeaem) [reply src/smos/smos_l4/reshuffle.py]: done

File: src/smos/smos_l4/reshuffle.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    if val in ['True', 'true', 't', 'T', '1']:
+        return True
+    else:
+        return False
+
+
+def reshuffle(input_root, outputpath,
```
[the remark this answers is not in the document]
Comment (Adeaem) [reply src/smos/smos_l4/reshuffle.py]: done

File: src/smos/smos_l4/interface.py (hunk 2 above)
Comment (Adeaem) [reply src/smos/smos_l4/interface.py]: agree. other review requests I completed I believe. 
merge and make this and issue?

File: src/smos/smos_l4/interface_l4.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+
+import warnings
+import numpy as np
+import os
+from pygeobase.io_base import ImageBase, MultiTemporalImageBase
+from pygeobase.object_base import Image
+from datetime import timedelta, datetime
+from netCDF4 import Dataset, date2num, num2date
+from smos.grid import EASE25CellGrid
+from smos.smos_ic import dist_name as subdist_name
```
Comment (wpreimes) [inline src/smos/smos_l4/interface_l4.py:33]: Move to general part of the package and import from there?

File: src/smos/smos_l4/interface_l4.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+        self.img = None  # to be loaded
+        self.glob_attrs = None
+
+        self.float_fillval = float_fillval
+
+    def get_global_attrs(self, exclude=('history', 'NCO', 'netcdf_version_id', 'contact',
+                                        'institution', 'creation_time')):
+        return {k: v for k, v in self.glob_attrs.items() if k not in exclude}
+
+    def _read_empty(self):
```
Comment (wpreimes) [inline src/smos/smos_l4/interface_l4.py:102]: Same for IC and L4, should be part of the base class

File: src/smos/smos_l4/interface_l4.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+
+        self.grid = grid
+
+        self.image_missing = False
+        self.img = None  # to be loaded
+        self.glob_attrs = None
+
+        self.float_fillval = float_fillval
+
+    def get_global_attrs(self, exclude=('history', 'NCO', 'netcdf_version_id', 'contact',
```
Comment (wpreimes) [inline src/smos/smos_l4/interface_l4.py:98]: Same for IC and L4, should be part of a common base class

File: src/smos/smos_l4/interface_l4.py [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+                return_img[key] = np.flipud(return_img[key].reshape(rows, cols))
+
+            self.img = Image(self.grid.activearrlon.reshape(rows, cols),
+                             np.flipud(self.grid.activearrlat.reshape(rows, cols)),
+                             return_img,
+                             return_metadata,
+                             timestamp)
+        return self.img
+
+    def write(self, image, **kwargs):
```
Comment (wpreimes) [inline src/smos/smos_l4/interface_l4.py:243]: almost the same as for IC, should be part of the common base class

**Review by wpreimes — CHANGES_REQUESTED** [review]
There is still too much duplication in my opinion, the 2 interface files are almost identical. We should make one base class and inherit 2 child classes (IC and L4) from that.

**Comment by wpreimes** [comment]
> 
> 
> There is still too much duplication in my opinion, the 2 interface files are almost identical. We should make one base class and inherit 2 child classes (IC and L4) from that.

Compare here @Adeaem https://github.com/TUW-GEO/ecmwf_models/tree/master/src/ecmwf_models

**Pull request merged**